src/errors.rs

```diff
@@ -1,16 +1,125 @@
 //! Error management module
 
-use crate::encoding::Decoder;
+use crate::encoding::{Decoder, EncodingError};
 use crate::escape::EscapeError;
 use crate::events::attributes::AttrError;
-use crate::name::QName;
-use crate::utils::write_byte_string;
+use crate::name::{NamespaceError, QName};
 use std::fmt;
 use std::io::Error as IoError;
-use std::str::Utf8Error;
-use std::string::FromUtf8Error;
 use std::sync::Arc;
 
+/// Combines errors into one error enum where `Error::source`, `Display` and `From<SourceError>` is already defined
+///
+/// **Usage**
+/// ```markdown
+/// combined_error! {
+///     /// ANY DOCS HERE
+///     #[ANY_ATTRIBUTE]
+///     NAME_OF_ERROR_TYPE ";"
+///     ERROR_VARIANT "(" ERROR_TYPE ?STORED_TYPE ")" FORMAT_STRING ","
+/// }
+/// ```
+/// - Docs and attributes are optional and apply to the new error type
+/// - FORMAT_STRING: Format string passed directly to `write!` (the error is the only parameter passed to `write!`)
+/// - STORED_TYPE: Optional inner type (like Arc<ERROR_TYPE>)
+///   - where: ERROR_TYPE: Into<STORED_TYPE>
+///
+/// Additionally, all source errors are required to have a `Into<crate::Error>`
+/// implementation so that all errors can be converted to `crate::Error`
+macro_rules! combined_error {
+    (@choose ($a:tt) ($b:tt)) => { $b };
+    (@choose ($a:tt)) => { $a };
+    (
+        $(#[$($attrss:tt)*])*
+        $name:ident;
+        $($variant:ident($error:path $(, $inner_type:path)?) $fmt_str:literal),+ $(,)?
+    ) => {
+        #[derive(Debug)]
+        #[allow(missing_docs)]
```

Comment:
I would like to update macro and write a documentation for each generated variant under which circumstances it will be returned. That is not always obvious from the name and description of the inner error.
```suggestion
```
Also, I would prefer to have more traditional syntax for defining enums, in order to in the end the written code will look like:
```rust
combined_error! {
  /// Doc
  // derives
  pub enum SpecificError {
    /// Variant 1 doc
    Variant1(Variant1Error) => "display 1 text",
    /// Variant 2 doc
    Variant2(Variant2Error) => "display 2 text",
  }
}
```
Please also derive `Debug` for each error type.